### ISSUE TITLE
Implement a MatrixMultiply layer

### ISSUE THREAD

**Issue by mobiusinversion** [keras-team/keras#11158]
As far as I can tell from documentation, Keras does not have a MatrixMultiplication layer, only Dense, TimeDistributed and the dot function.  Is that correct?

If so, this feature request is to add `layers.MatrixMultiply()`.  Naturally one can implement this using a custom layer and dot.  But why would this be required for the general population?

The calling semantics would be something like:

    layers.MatrixMultiply((l, m))(input) # input is a batch of (m x n) tensors.

**Comment by gabrieldemarmiesse** [comment]
I don't think the use case is common enough to go in core keras. You can easily implement this with a Lambda layer and K.batch_dot() in one line. Since this is simple to implement, and that we already have a Dense layer, I don't think it's worth increasing the maintenance burden of keras.

**Comment by mobiusinversion** [comment]
@gabrieldemarmiesse, that's why I mentioned it was straight forward to implement, this was just a user friendlyness issue.  I'm not sure any of us knows the commonality of the use case.  Thanks for your viewpoint.

**Comment by mobiusinversion** [comment]
Hey @fchollet, do you think this that uncommon?

**Comment by mobiusinversion** [comment]
@gabrieldemarmiesse, I am not sure this is a one liner if you want the matrix used for multiplication to be trainable.  Could you please provide an example if so?

**Comment by mobiusinversion** [comment]
@gabrieldemarmiesse @fchollet I am not sure you should dismiss this so quickly and outright.  It is overly complex to design simple operations like Luong based attention in Keras.  For example, from here:

        self.attention_transposed_rnn_outputs = layers.Permute((2, 1))(self.encoder_rnn_outputs)

        self.attention_encoder_matrix_multiply = layers.TimeDistributed(
            layers.Dense(self.rnn_size, use_bias=False))(self.attention_transposed_rnn_outputs)

One would now have to craft boilerplate around TimeDistributed matrix multiplication over decoder hidden states, over batch sizes.  If this is an obscure use case, then that tells me Keras is meant for basic networks, but we know that's not the case and the functional API is complete.  You might want to consider that this is an annoying use case that could be a better user experience.

**Comment by mobiusinversion** [comment]
@gabrieldemarmiesse Also, I cannot used custom Lambda layers because of serialization constraints.  This will need to go into protocol buffers.  You see this is not entirely trivial, so I would appreciate if you could reopen this, at least for comments on alternatives.

**Comment by gabrieldemarmiesse** [comment]
Sure, I can reopen, I seems that not everything is clear. There are several things to discuss then.

* What are the serialisation constraints that you are mentioning? If it doesn't work well, it's important to have feedback. 

* Could you implement this custom layer and put your implementation in this issue? It seems that I didn't fully understand what type of layer you were talking about. Usually, code is a good way to make sure that everbody understands the same thing. 

Thank you for your help :)

**Comment by mobiusinversion** [comment]
Hi @gabrieldemarmiesse, thank you for the opportunity to discuss and gather wisdom from the experts :)  Regarding the items for discussion you mentioned, 

- Serialization: In order to make Keras models usable in a high speed, low latency web serving setting,  one of the most efficient and effective paths is to serialize the underlying TensorFlow model down to the protocol buffers.  It is not possible to use h5 directly since the only existing library I am aware of which provides the corresponding Java glue is DL4J, and that implementation has issue after issue and has not proven stable enough for real world usage.  Besides, I don't think they will be able to support Lambda layers either.  So the Serialization proceeds in the canonical way:

      backend.set_learning_phase(0)
      sess = backend.get_session()
      builder = tf.saved_model.builder.SavedModelBuilder("./model")
      builder.add_meta_graph_and_variables(sess, [tf.saved_model.tag_constants.SERVING])
      builder.save(False) 

This will however not respect any kind of logic present in a Lambda layer as it is outside the TensorFlow metagraph.

- Regarding a working example, I am still working through the details.  When I wrote this code snippet:

      self.attention_encoder_matrix_multiply = layers.TimeDistributed(
          layers.Dense(self.rnn_size, use_bias=False))(self.attention_transposed_rnn_outputs)

The intent was to compute `W_a * h_s` as in Luong Attention.  The next step would be to left dot the decoder hidden state, i.e. `h_t * W_a * h_s`, where h_s is the transposed matrix of hidden encoder states and `h_t` is an individual hidden state.  

![scoring_functions](https://user-images.githubusercontent.com/955489/45626090-6b441800-ba43-11e8-92fc-334338d02463.png)

However, now looking at the above code, I think it might actually compute `h_s * W_a`, but I need to confirm.  I will work more on this during the week (I lead a Keras / TensorFlow study group).  It would naturally be quite transparent to do all this directly in TensorFlow, but Keras is just so concise and elegant that it would be a shame not to give it a shot first.  Please stay tuned and let me know if you have any ideas in the meantime.  Thanks!

**Comment by gabrieldemarmiesse** [comment]
> This will however not respect any kind of logic present in a Lambda layer as it is outside the TensorFlow metagraph.

I'm not familiar with Tensorflow's metagraph. Is this issue also present with keras core layers and keras custom (user-made) layers too?

>  It would naturally be quite transparent to do all this directly in TensorFlow

If you know how to do it in Tensorflow, then you should be able to find the equivalent keras operations from this file: https://github.com/keras-team/keras/blob/master/keras/backend/tensorflow_backend.py.

You can then check if it correspond to what already exist. Let's say for example the Dense layer here: https://github.com/keras-team/keras/blob/ef0c95aca2fd19666711422c14a6b16f8025ed84/keras/layers/core.py#L767

An implementation (tensorflow or keras or pytorch or whatever) would help a lot to know exactly as you're looking for since formulas and papers often.... well, let's say the devil's in the details.

**Comment by farizrahman4u** [comment]
@mobiusinversion 

* I believe the lambda layer approach will work and any ops called in your lambda layer logic will be serialized to the protobuff (if it doesn't that's a bug of its own, but I think it does)

* You dont have to use the `TimeDistributed` wrapper in conjunction with `Dense`, `Dense` can handle sequences on its own.

* Back to your original issue:

```python
layers.MatrixMultiply((l, m))(input) # input is a batch of (m x n) tensors
```

This can be achieved via existing Keras layers using:

```python
input_transposed = Permute((2, 1))(input)
dense = Dense(l, use_bias=False)
output = dense(input_transposed)
output = Permute((2, 1))(output)
```

**Comment by mobiusinversion** [comment]
@farizrahman4u, would this cover a left hand side matrix multiplication?  That is, where the weights are on the left hand side of the matmul:  W * X?  This should not be this hard to do in Keras.  I'd be interested to hear the wisdom of @fchollet on this one.

**Comment by farizrahman4u** [comment]
In symbolic graphs, any python based logic will be reflected in the way nodes are connected to each other, or will create ops (for e.g pythonic slicing x[:, :dim] will create a SliceOp etc), and hence can be carried over over to C++/Java via protobuff.

**Comment by dwilliams999** [comment]
For anyone who comes across this issue @farizrahman4u @gabrieldemarmiesse here is a pure Keras + tf solution to right side and left side matmul layers which broadcast over batches:

    class RightSideMatMul(Layer):
    
        def __init__(self, shape=None, dtype=np.float32, **kwargs):
            self.rows = shape[0]
            self.columns = shape[1]
            self.kernel = None
            self.dtype = dtype
            super(RightSideMatMul, self).__init__(**kwargs)
    
        def build(self, input_shape):
            self.kernel = self.add_weight(
                name='kernel',
                shape=(self.rows, self.columns),
                initializer='glorot_uniform',
                dtype=self.dtype,
                trainable=True)
            super(RightSideMatMul, self).build(input_shape)
    
        def call(self, x, **kwargs):
            x_reshaped = tf.reshape(x, [-1, self.rows])
            x_multiplied = tf.matmul(x_reshaped, self.kernel)
            return tf.reshape(x_multiplied, [-1, x.shape[1], self.columns])
    
    
    class LeftSideMatMul(Layer):
    
        def __init__(self, shape=None, dtype=np.float32, **kwargs):
            self.rows = shape[0]
            self.columns = shape[1]
            self.kernel = None
            self.dtype = dtype
            super(LeftSideMatMul, self).__init__(**kwargs)
    
        def build(self, input_shape):
            self.kernel = self.add_weight(
                name='kernel',
                shape=(self.rows, self.columns),
                initializer='glorot_uniform',
                dtype=self.dtype,
                trainable=True)
            super(LeftSideMatMul, self).build(input_shape)
    
        def call(self, x, **kwargs):
            kernel_transpose = tf.matrix_transpose(self.kernel)
            x_transpose = tf.matrix_transpose(x)
            matmul = tf.matmul(tf.reshape(x_transpose, [-1, self.columns]), kernel_transpose)
            matmul_reshaped = tf.reshape(matmul, [-1, x.shape[2], self.rows])
            return tf.matrix_transpose(matmul_reshaped)
